### ISSUE TITLE
Integration with the Geant4 Tasking

### ISSUE THREAD

**Issue by whokion** [celeritas-project/celeritas#71]
As Geant4 provides the C++11 tasking capability since the 10.7.beta release, demonstrate a hybrid (concurrent) computing application that offloads EM particle transport on GPU using celeritas and simulates the rest of simulation (i.e., hadronic physics processes) on the host CPU for a typical HEP detector simulation at LHC:
- build a generic HEP application with the Geant4 tasking manager and connect specialized tasks to celeritas (or vice versa)
- collect EM particles from the Geant4 tracking loop (mostly photons and electrons from hadronic interactions) 
- trigger celeritas kernels for specialized GPU tasks for EM particles 
- merge sensitive hits from Geant4 and celeritas 
- optimize the load-balancing between the CPU host and devices and evaluate performance

**Comment by drbenmorgan** [comment]
Hi @whokion, a quick ping to say I'd be happy to work on this task. Let me know how I can help/collaborate!

**Comment by whokion** [comment]
@drbenmorgan, sounds great!  We may discuss a potential path how to integrate celeritas GPU EM tasks with the Geant4 task model later, I would like to propose following steps: 
1) create a directory, something like "geant4-tasking" under celeritas/app 
2) add a Geant4 application for generic HEP applications (something like a standalone application with cms gdml and a magnetic field map, but extendable with other detector geometry configurations) simulation typical HEP (pythia) events.
3) collect EM (e+,e-, gamma) particles from the tracking loop and store minimal information into celeritas particle stacks (which will be sent to GPU by chunks) and kill those Geant4 EM tracks on GPU
4) launch kernels (for now, we just may send stacked data from H2D until the celeritas EM transportation is ready)    
5) do steps 3 and 4 concurrently with the Geant4 task model (a task group process multiple Geant4 events, aka Geant4 MT, and another task group to handle multiple EM kernels) 

I can quickly prepare 1-4 steps as a starting point as I already exercised those steps.  The key task is the step 5 when the Geant4 task is readily to be tested.  Of course, we may get helps from Johnathan as needed, but can play many other things with 1-4 even before the Geant4 task is fully functional.  Please let me know if you have different ideas.

**Comment by drbenmorgan** [comment]
@whokion, that sounds like a good plan, so if you want to get that directory and initial code in place, we can iterate from there. Geant4 10.7 is out tomorrow, and I'll get the spack package updated so we're good to go with the tasking as and when we want to.

**Comment by whokion** [comment]
@drbenmorgan okay, will do and get back to you once it is ready to be reviewed - also keep @sethrj in the loop.  Thanks for your contribution.

**Comment by sethrj** [comment]
This is great, and I'm glad you guys are making headway on this! If "Celeritas library as a Geant4 toolkit app" has the potential to grow beyond an experimental stage, or if you want more granularity in the development process (fewer reviews, looser style guides), may I suggest the possibility of creating a new repository under the celeritas project? A potential disadvantage is that might decouple your development work from the (perhaps) rapid changes that might be going on in the main repo. Can we three meet briefly (maybe Friday tomorrow) to discuss the best strategy moving forward?

**Comment by drbenmorgan** [comment]
Hi @sethrj, @whokion, a quick chat would be useful to strategise. Am free today up to 1pm CST, so any time before that would be fine.

**Comment by whokion** [comment]
@sethrj, @drbenmorgan How about 11am (5pm@UK) at the usual celeritas zoom room?

**Comment by drbenmorgan** [comment]
@whokion, @sethrj, 11am (5pm UK) works for me.

**Comment by whokion** [comment]
@drbenmorgan Very good. @sethrj 12:00 your time.  If no conflict, let's meet at
https://exascaleproject.zoomgov.com/j/1610125573?pwd=akJib2VCMUQxS2pTcnEyOHpWcldIZz09

**Comment by sethrj** [comment]
Per our discussion we're breaking off the Geant4 tasking integration into a separate repo that will integrate with Celeritas,  https://github.com/celeritas-project/acceleritas , so I'm going to close this issue.
